### PR TITLE
Retry downloading packages on failure

### PR DESCRIPTION
**Pull request by pzakha**
This is a proposed workaround for #380 

## Testing
relevant sub-job from ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/pre-push/7356/consoleFull
ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2506/

### CODE REVIEW

**Comment by prakashsurya** [comment]
Can you please add information to help explain why we'd want to implement this suggested solution, vs. other alternative solutions? Can you also add information about any alternative solutions that were investigated (if any), and the benefits/drawbacks of these alternatives when compared to what you're proposing here.

**Comment by prakashsurya** [comment]
Can you also add information to help explain why the download failures occur? I think we should fully understand why the failures occur, especially given that we're using infrastructure that's under our control, before we attempt to workaround the failures.

**Comment by pzakha** [comment]
Most of the failures in #380 are in the stage where we download packages.

We've spend some time investigating why the failures are occurring but did not have much luck so far. I think we should spend more time root causing this issue after we ship 6.0.

In the meanwhile, this can help with:
a) Reducing the number of failures and so reducing the burden to monitor failed builds.
b) If this does not reduce failures and we see the download of a package fail for several attempts in a row, this could give us more insight into the problem.

**Comment by prakashsurya** [comment]
Can you also respond to my first comment, about potential alternative solutions, and why you believe this is the best path to take.

**Comment by pzakha** [comment]
A few alternatives:
 1. Re-use packages downloaded previously during the build, so "apt download" can get them from cache. This may not work for all packages, but should work for most.
2. Spend more time investigating the root cause for the failure. One idea was to monitor the workload on the server.

Both alternatives would most likely take a good amount of time, and would probably have to be done after we ship 6.0.

**Comment by pzakha** [comment]
I think implementing this workaround now will help us to get us a better success rate on our builds.

**Comment by prakashsurya** [comment]
Why is shipping 6.0 a factor here? Is this blocking the release?

Unless I'm mistaken, the build is currently flaky, but we can re-run it when it fails, and continue to re-run until we get a successful build right? and then we use that successful build to ship the release?

It's not clear why we'd actively pursue a short term solution due to the 6.0 release, even though this issue isn't blocking the release?

**Comment by pzakha** [comment]
The main idea here is to reduce the maintenance burden of monitoring failed builds, so that we can focus on real failures instead of infrastructure flakiness. If we think dealing with flakiness is fine for the next few weeks, until we get more time to root cause this, I'm fine with dropping this.

File: live-build/config/hooks/configuration/download-packages.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+PACKAGES=()
+while read -r package; do
+	PACKAGES+=("$package")
+	if [[ "${#PACKAGES[@]}" -ge 20 ]]; then
+		download_packages "${PACKAGES[@]}"
+		PACKAGES=()
+	fi
+done </tmp/packages
+
+[[ "${#PACKAGES[@]}" -ne 0 ]] && download_packages "${PACKAGES[@]}"
```
Comment (prakashsurya) [inline live-build/config/hooks/configuration/download-packages.sh]: could this be replaced with something like this:
```
export -f download_packages
dpkg-query -Wf '${Package}=${Version}\n' | xargs -n 20 bash -c 'download_packages "$@"' _
```
?

Comment (pzakha) [reply live-build/config/hooks/configuration/download-packages.sh]: That's much more concise. It seems like it doesn't abort if `download_packages` fails though, not sure why.

Comment (prakashsurya) [reply live-build/config/hooks/configuration/download-packages.sh]: What do you mean? Here's what I see in my testing...

- All succeed, it returns success:
```
± function dostuff() { exit 0; }
± echo 1 2 3 4 | xargs -n2 bash -c 'dostuff "$@"' _
± echo $?
0
```

- All fail, it returns failure:
```
± function dostuff() { exit 1; }
± echo 1 2 3 4 | xargs -n2 bash -c 'dostuff "$@"' _
± echo $?
123
```

- one fails, it returns failure:
```
± function dostuff() { local R=$((RANDOM % 2)); echo $R; exit $R; }
± echo 1 2 3 4 | xargs -n1 bash -c 'dostuff "$@"' _
0
1
0
0
± echo $?
123
```

File: live-build/config/hooks/configuration/download-packages.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+function download_packages() {
+	echo "Downloading: $*"
+	for _ in {1..5}; do
+		apt-get download "$@" && return
+		echo "Download failed, retrying."
+		sleep 5
+	done
+	echo "Failed to download packages after 5 attempts. Aborting." >&2
+	exit 1
```
Comment (prakashsurya) [inline live-build/config/hooks/configuration/download-packages.sh]: we use `exit` here, but `return` earlier in this function.. we should try to be consistent and either use `exit` for both, or `return` for both.

File: live-build/config/hooks/configuration/download-packages.sh (hunk 1 above)
Comment (pzakha) [reply live-build/config/hooks/configuration/download-packages.sh]: Yeah, in my quick testing I was running something very similar to your 3rd case, where we can see it doesn't stop on failure. It does return an error code at the end though, so it should be fine. I've updated the review.

**Comment by pzakha** [comment]
re-testing ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2507/

**Comment by pzakha** [comment]
@prakashsurya I think I've addressed your feedback, can you take another look?

**Comment by prakashsurya** [comment]
As others have asked offline, how are we going to monitor and/or track the effectiveness of the retry loop? e.g. how will we track how many times we have to retry? how many retries are needed until it works? which packages required and retry, and which dont? how do we intend to use the data we get from this retry loop to help address the flakiness of the package server?

Also, if this is a short term solution, what criteria will be using to know when we should remove the retry loop, in favor of whatever long term solution we adopt?

I still feel like rushing into a short term solution might not be the right approach, since if this does alleviate build failures, I'd bet this will become the long term solution (e.g. due to other work taking priority). I'd like to get others to weigh in before approving this.

Tangentially, who is in charge of maintaining the package server? What kind of SLA does the server have? Is this flakiness expected? Have we opened a dialog/ticket with the folks that maintain the service?

**Comment by sebroy** [comment]
@prakashsurya I filed internal tool issue TOOL-8699 against the internal package mirror as a result of these failures.

**Comment by prakashsurya** [comment]
@sebroy Thanks! Are you good with this approach?

**Comment by sebroy** [comment]
@prakashsurya I'm fine with this approach, although to your point, it would be great to include whether retries were needed, against what package, and how many, as part of the report of the job that's included in our slack notifications (e.g. similar to how we include some details in the `linux-pkg-update` job) so that I or the monitor can identify a pattern with little effort.

**Comment by pzakha** [comment]
> I still feel like rushing into a short term solution might not be the right approach, since if this does alleviate build failures, I'd bet this will become the long term solution (e.g. due to other work taking priority).

I agree there's definitely a risk of this becoming a long term solution if things work. I think if this workaround actually fixes the issue, then it's not a big deal that we just leave it in place. The bigger risk is that we'd just mask the issue temporarily, and it would come back later and fail even with the workaround in place.

> it would be great to include whether retries were needed, against what package

That would definitely require quite a bit more work. If we want to go this route, I'd propose we hold off on this for now.

**Comment by sebroy** [comment]
@pzakha would there be another way to fetch and report that same information, perhaps via post-processing? I'd like to maybe run such a report every week to see what the state of the job is, whether it had to retry the package updates, etc. etc. without having to spend a lot of time clicking through all of the successful runs.

**Comment by pzakha** [comment]
There's definitely ways of doing that via post-processing, such as parsing the output of each `appliance-build-stage1` job and looking for specific strings. This may put quite some load on the Jenkins server though as the failure rate is pretty low and the script will have to parse several hundreds of jobs each week.

**Comment by pzakha** [comment]
Superseded by https://github.com/delphix/appliance-build/pull/415

**Pull request closed without merging**